File: leetcode/Python/74. Search a 2D Matrix.py
from bisect import bisect_left 

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        '''
        How to solve: 답안 참고
        접근 방식은 동일하나 내 접근 방식에서 코드 구현이 미흡했음. 이진 탐색에 대한 이해도가 낮음.
        
        Runtime: 66 ms, faster than 46.05% of Python3 online submissions for Search a 2D Matrix.
        Memory Usage: 14.3 MB, less than 88.83% of Python3 online submissions for Search a 2D Matrix.
        
        '''
        r = bisect_left(matrix, target, key=lambda row: row[-1])  # or key=itemgetter(-1)
        return r < len(matrix) and matrix[r][bisect_left(matrix[r], target)] == target
            
        
#         m = len(matrix)
#         n = len(matrix[0])
        
#         first_entrance = []
#         for i in range(m):
#             first_entrance.append(matrix[i][0])
#         print("first_entrance is: ", first_entrance)
        
#         first_index = bisect_left(first_entrance, target) if (bisect_left(first_entrance, target) == target) else (bisect_left(first_entrance, target)-1)
#         print("first index: ", first_index)
#         if first_index >= m or first_index < 0:
#             return False
        
#         second_index = bisect_left(matrix[first_index], target)
#         if second_index >= m or second_index < 0:
#             return False
#         print("second index: ", second_index)
#         print("target is: ", matrix[first_index][second_index])
#         if matrix[first_index][second_index] == target:
#             return True
#         return False
        

File: leetcode/Java/74. Search a 2D Matrix.java
public class 74. Search a 2D Matrix {

    /*
     * Runtime: 0 ms, faster than 100.00% of Java online submissions for Search a 2D Matrix.
     * Memory Usage: 42.7 MB, less than 53.51% of Java online submissions for Search a 2D Matrix.
     * 시간복잡도: O(logMN) -> 바이너리 서치 이용 & 전체 길이가 M*N인 하나의 sorted List에서 바이너리 서치를 한다고 생각하면 된다.
     * 공간 복잡도: O(1) -> 데이터 양에 따른 공간 증가가 없음. 상수.
     */
    class Solution {
        public boolean searchMatrix(int[][] matrix, int target) {
            // Exception case 정리
            if (matrix == null || matrix.length == 0) {
                return false;
            }
            // 값 설정
             
            int rows = matrix.length, cols = matrix[0].length;
            int start = 0;
            int end = rows * cols -1 ;
            
            while (start <= end) {
                int mid = (start + end) /2;
                // true case
                if (matrix[mid / cols][mid % cols] == target) {
                    return true;
                }
                if (matrix[mid / cols][mid % cols] < target) { // mid보다 target이 크다 -> start를 올려줘야 한다.
                    start = mid + 1;
                } else {
                    end = mid -1;
                }
            }
            return false;
        }
    }
    
}
